tests/python/test_python_frontend.py

```diff
@@ -4458,6 +4457,86 @@ def fusion_func(fd: FusionDefinition):
         for i in range(num_out):
             self.assertEqual(nvf_out[i].data_ptr(), inputs[0].data_ptr())
 
+    @unittest.skipIf(is_pre_ampere(), "Only supported on Ampere and newer devices.")
+    def test_sdpa_fwd(self):
+        def fusion_func(
+            fd: FusionDefinition, has_dropout: bool, has_causal: bool, has_scale: bool
+        ):
+            q = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            k = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            v = fd.define_tensor(
+                shape=[-1, -1, -1, -1],
+                contiguity=[True, True, True, True],
+                dtype=DataType.BFloat16,
+                is_cpu=False,
+            )
+            dropout_p, is_causal, scale = None, None, None
+            if has_dropout:
+                dropout_p = fd.define_scalar(None, dtype=DataType.Double)
+            if has_causal:
+                is_causal = fd.define_scalar(None, dtype=DataType.Bool)
+            if has_scale:
+                scale = fd.define_scalar(None, dtype=DataType.Double)
+            attn, *intermediate_results = fd.ops.sdpfa_fwd(
+                q, k, v, dropout_p, is_causal, scale
+            )
+            fd.add_output(attn)
+
+        N, H, L, S, E = 4, 8, 16, 16, 8
+        qkv = [
+            torch.randn((N, H, L, E), dtype=torch.bfloat16, device="cuda:0"),
+            torch.randn((N, H, S, E), dtype=torch.bfloat16, device="cuda:0"),
+            torch.randn((N, H, S, E), dtype=torch.bfloat16, device="cuda:0"),
+        ]
+
+        dropout_vals = [None, 0.0, 0.2]
+        is_causal_vals = [None, True, False]
+        scale_vals = [None, 1 / E**0.5, 1e-3]
+        torch.manual_seed(0)
+        # TODO: Try to move this to pytest_ops.py. Currently, it does not work since the API between nvFuser and torch differs.
+        for dropout_p, is_causal, scale in itertools.product(
+            dropout_vals, is_causal_vals, scale_vals
+        ):
+            with self.subTest(dropout_p=dropout_p, is_causal=is_causal, scale=scale):
+                # Reset the FusionCache or the fusion would not recompile for all subtests, failing checks in exec_nvfuser.
+                FusionCache.reset()
+                has_dropout = True if dropout_p is not None else False
+                has_causal = True if is_causal is not None else False
+                has_scale = True if scale is not None else False
+                inputs = [*qkv]
+                for param in [dropout_p, is_causal, scale]:
+                    if param is not None:
+                        inputs.append(param)
+                with torch.random.fork_rng(range(torch.cuda.device_count())):
```

Comment:
According to https://pytorch.org/docs/stable/random.html#torch.random.fork_rng, fork_rng by default operates on all devices, so the `range(...)` part is unnecessary.

Reply:
That's strange. I got an error that since there are multiple devices, I need to specify an ID.

Reply:
The doc says
> will emit a warning if your machine has a lot of devices, since this function will run very slowly in that case.
 
Could that be the reason? I've no idea what "a lot" means. 

Anyhow, this test doesn't use multiple GPUs, so it should be enough to specify just torch.cuda.current_device().